net/download/service/src/main/java/de/danoeh/antennapod/net/download/service/feed/FeedUpdateWorker.java

```diff
@@ -127,36 +130,56 @@ private Notification createNotification(@Nullable List<Feed> toUpdate) {
                 .build();
     }
 
+    private void updateNotification(List<Feed> toUpdate) {
+        if (ContextCompat.checkSelfPermission(getApplicationContext(), Manifest.permission.POST_NOTIFICATIONS)
+                == PackageManager.PERMISSION_GRANTED) {
+            notificationManager.notify(R.id.notification_updating_feeds, createNotification(toUpdate));
+        }
+    }
+
     @NonNull
     @Override
     public ListenableFuture getForegroundInfoAsync() {
         return Futures.immediateFuture(new ForegroundInfo(R.id.notification_updating_feeds, createNotification(null)));
     }
 
     private void refreshFeeds(List<Feed> toUpdate, boolean force) {
-        while (!toUpdate.isEmpty()) {
-            if (isStopped()) {
-                return;
-            }
-            if (ContextCompat.checkSelfPermission(getApplicationContext(), Manifest.permission.POST_NOTIFICATIONS)
-                    == PackageManager.PERMISSION_GRANTED) {
-                notificationManager.notify(R.id.notification_updating_feeds, createNotification(toUpdate));
-            }
-            Feed feed = toUpdate.get(0);
-            try {
-                if (feed.isLocalFeed()) {
-                    LocalFeedUpdater.updateFeed(feed, getApplicationContext(), null);
-                } else {
-                    refreshFeed(feed, force);
+        //Unchecked is used to prevent a race condition where a feed finishes downloading before
+        // we're done submitting all of the jobs
+        List<Feed> unchecked = new ArrayList<Feed>(toUpdate);
```

Comment:
Is this copy actually needed? You are only updating the list in a synchronized block anyway. Is this maybe a leftover from an earlier experiment?

Reply:
I think it might be, because we are iterating over `toUpdate` and I guess we shouldn't remove items from a list we are currently iterating over.

Reply:
Ah, right

Reply:
Yeah, that's correct if it were to somehow managed to get to the remove before the for loop was done, we
d get an ConcurrentModificationException so the copy is there to prevent that.